uwblueprint/project-read-frontend
# Implement multiple select questions

### Notion ticket link



[checkbox questions ☑️ ](https://www.notion.so/uwblueprintexecs/Developer-Hub-30e9b6c82af24054b31acf7c5e822c89?p=c76204f80c2e4dcfa1e8bca175be023b)



### Implementation description

this is a complex but important PR! please review & test with care 💙  

this PR implements multiple select questions by introducing a new way to interpret the values in students' `information` data. if the value string contains the `\n` character, in the multi-select component, we consider this to be a separator and turn this string into an array of strings. none of our dynamic fields support multi-line inputs (and won't for the foreseeable future), so this character can never be entered by a user manually.

<img width="415" alt="Screen Shot 2021-08-10 at 10 44 47 PM" src="https://user-images.githubusercontent.com/37346399/128961562-0a3994e0-3cfc-441c-b087-64c668f12edd.png">

- implemented a MultipleSelect component that considers the `\n` character to be a separator in the value string
  - transforms the string value into an array: `"dogs\nfrogs"` → `[dogs, frogs]`
  - when a value is selected, call `onChange` after transforming the array back into a string: `[dogs, frogs]` → `"dogs\nfrogs"`
  - wrote tests to ensure i got this right 😅 
- added logic in the `Field` component to display values when the question type/options have changed
  - if the value is not one of the field's options, display it as one
  - text/select to multiselect: display the value as is
  - multiselect to text/select: change the value to a comma-separated string (`"dogs\nfrogs"` → `"dogs, frogs"`)
- some styling



### Steps to test

1. run the backend on uwblueprint/project-read-backend#109, and run migrate & load_initial_data
1. yarn start
1. yarn run test
1. open the sidebar and registration form, and play with the multi-select & saving data (dietary restrictions or technology access questions from dummy data)
1. change question types around in the django admin, and verify that the inputs/values on the frontend display & behave as expected



### What should reviewers focus on?

- does this approach sound okay?
- is there another character delimiter i should use?

### Checklist
